Odata-docs/webapi-8/fundamentals/property-routing.md

```diff
@@ -0,0 +1,976 @@
+
+The following request returns the raw value of the number of items in the `ContactPhones` collection-valued property on customer 1:
+```http
+GET http://localhost:5000/odata/Customers(1)/ContactPhones/$count
+```
+
+For the above request to be conventionally-routed, a controller action named `GetContactPhones` that accepts the key parameter is expected. The action should be decorated with `EnableQuery` attribute:
+```csharp
+[EnableQuery]
+public ActionResult GetContactPhones([FromRoute] int key)
+{
+    var customer = customers.SingleOrDefault(d => d.Id.Equals(key));
+
+    if (customer == null)
+    {
+        return NotFound();
+    }
+
+    return Ok(customer.ContactPhones);
+}
+```
+
+The expected response is shown below:
+```console
+1
+```
+
+### Add an item to a collection-valued property
+The route templates for this request are:
+- `POST ~/{entityset}({key})/{collectionvaluedproperty}`
+- `POST ~/{entityset}/{key}/{collectionvaluedproperty}`
+- `POST ~/{singleton}/{collectionvaluedproperty}`
+
+The following `POST` request adds an item to the `ContactPhones` primitive collection-valued property on customer 1:
+```http
+POST http://localhost:5000/odata/Customers(1)/ContactPhones
+```
+
+Here's the request body:
+```json
+{
+    "value": "798-507-2014"
+}
+```
+
+The payload for adding an item to a primitive collection-valued property should be a JSON object containing a single property named `value`.
+
+For the above request to be conventionally-routed, a controller action named `PostToContactPhones` is expected. The controller action should accept two parameters - the first is the key parameter and the second a parameter of type `string` decorated with `FromBody` attribute:
+```csharp
+public ActionResult PostToContactPhones([FromRoute] int key, [FromBody] string contactPhone)
+{
+    var customer = customers.SingleOrDefault(d => d.Id.Equals(key));
+
+    if (customer == null)
+    {
+        return NotFound();
+    }
+
+    customer.ContactPhones.Add(contactPhone);
+
+    return Accepted();
+}
+```
+
+The response status code should be `201`. You can query for the entity to confirm the result of the `POST` operation.
+
+### Adding an item to a collection-valued property declared on a derived type
+The route templates for this request are:
+- `POST ~/{entityset}({key})/{cast}/{collectionvaluedproperty}`
+- `POST ~/{entityset}/{key}/{cast}/{collectionvaluedproperty}`
+- `POST ~/{singleton}/{cast}/{collectionvaluedproperty}`
+
+The following `POST` request adds an item to the `ShippingAddresses` complex collection-valued property on enterprise customer 3:
+```http
+POST http://localhost:5000/odata/Customers(3)/PropertyRouting.Models.EnterpriseCustomer/ShippingAddresses
+```
+
+Here's the request body:
+```json
+{
+    "Street": "One Microsoft Way"
+}
+```
+
+Unlike adding an item to a primitive collection-valued property, the payload for adding an item to a complex collection-valued property is a JSON object of the complex type.
```

Comment:
@habbes It's a case of killing two birds using one stone. You realize that property routing routes templates are quite many and I had to be a bit frugal so the doc doesn't get even longer. To accomplish that, I added the primitive collection property to the base class and the complex collection property to the derived class - instead of having both category of properties on base and derived classes.

In the **Add an item to a collection-valued property** section I demonstrate 2 things:
 - Adding an item to a collection-valued property
 - Adding an item to a primitive collection-valued property

In the **Adding an item to a collection-valued property declared on a derived type** section I demonstrate 2 things:
- Adding an item to a collection-valued property declared on a derived type
- Adding an item to a complex collection-valued property

It's therefore right to say that **Adding an item to a collection-valued property declared on a derived type** section demonstrates _adding to a collection property_ and _adding to collection of complex types_, while **Add an item to a collection-valued property** section demonstrates _adding to a collection property_ and _adding to collection of primitive types_. I added the **note** about how the payload should look like in both scenarios. Otherwise I'd have to accommodate the following breakdown:
- Adding an item to a primitive collection-valued property
- Adding an item to a complex collection-valued property
- Adding an item to a primitive collection-valued property declared on a derived type
- Adding an item to a complex collection-valued property declared on a derived type

Even in the **retrieving** section, I used **Retrieving a property** in general to represent:
- Retrieving a primitive single-valued property
- Retrieving a complex single-valued property
- Retrieving a primitive collection-valued property
- Retrieving a complex collection-valued property

It's because the semantics are the same. There's no big difference between `~/Customers(1)/Name` (where `Name` is a primitive single-valued property), and `~/Customers(1)/BillingAddress` (where `BillingAddress` is a complex single-valued property), and `~/Customers(1)/ContactPhones` (where `ContactPhones` is a primitive collection-valued property):
```csharp
public ActionResult GetName([FromRoute] int key)
{
    //
    return Ok(customer.Name);
}

public ActionResult GetBillingAddress([FromRoute] int key)
{
    //
    return Ok(customer.BillingAddress);
}

public ActionResult GetContactPhones([FromRoute] int key)
{
    //
    return Ok(customer.ContactPhones);
}
```

If I took the breakdown/expressive route, I'd have to equally break down **Retrieving a property declared on a derived type** into:
- Retrieving a primitive single-valued property declared on a derived type
- Retrieving a complex single-valued property declared on a derived type
- Retrieving a primitive collection-valued property declared on a derived type
- Retrieving a complex collection-valued property declared on a derived type

 Let me know whether that clarifies or whether you still feel strongly about being more expressive or other alternative suggestions that you may have on how we can provide the clarity.

Reply:
Thanks for the clarifications. Your approach makes more sense.